AzureAD/microsoft-authentication-library-common-for-android
# Melissaahn/yubi kit override integration

### Links to Dependent PRs
- `keyboard` config changes in Broker Manifest: https://github.com/AzureAD/ad-accounts-for-android/pull/1869
- `keyboard` config changes in MSAL Manifest: https://github.com/AzureAD/microsoft-authentication-library-for-android/pull/1618
- Adding YubiKit Version to versions.gradle in android complete: https://github.com/AzureAD/android-complete/pull/135
- Adding YubiKit Version to versions.gradle in ADAL: https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/1689

## The Problem
Two main issues were brought about at the start of this EO YubiKey support work:

- When the YubiKey was plugged in, the host app, common, and broker code reacted adversely. While trying to authenticate via WebView, the WebView would seem to refresh upon YubiKey plug in, and this would result in an authentication flow that didn't seem to complete despite providing a correct upn and certificate/password. Additionally, if the YubiKey was plugged in or removed while in AzureSample's fragments, the app would crash.
- [YubiKit for Android](https://github.com/Yubico/yubikit-android) has a minimum API level of 19. The current minimum level of Common is 16.

## Why This Was Happening
The first issue was due to Android recognizing a YubiKey device as a keyboard, which I suspect has to do with the [YubiKey's Touch Triggered OTP feature](https://developers.yubico.com/Developer_Program/Guides/Touch_triggered_OTP.html#:~:text=The%20YubiKey%20has%20an%20integrated,input%20field%20or%20command%20prompt.). Android sees an external keyboard being plugged in as a configuration change [`configChanges`](https://developer.android.com/guide/topics/manifest/activity-element#config). The default behavior for `configChanges` is to restart the current Activity. This behavior checks out with the observed WebView behavior. In the case of the host app crashing, the account Spinner was attempting to reference a null context before the current fragment could get reattached.

In the second issue, bumping up the `minSdkVersion` of Common needed to be avoided in order to still support our customers workings with a min API level less than 19. 

## Changes Made
For the first issue, in all Manifests with Activities that could be exposed to a plugged in YubiKey, `keyboard` and `keyboardHidden` need to be added to the `configChanges` for each activity listed. This overrides the default behavior and allows the activities themselves to handle the configuration change (in this case, we do nothing). 

For the second issue, the `minSdkVersion` issue is avoided by [overriding <uses-sdk> for yubico.yubikit.android](https://developer.android.com/studio/build/manage-manifests#override_uses-sdk_for_imported_libraries). The library will only be referenced within blocks that ensure the device's api level is at 19 or above. 

~~Most of the smartcard CBA handling will be contained within a new `IChallengeHandler` `EoClientCertAuthChallengeHandler`. At the moment, the handler still handles on-device CBA the same (it extends `ClientCertAuthChallengeHandler`),~~ `ClientCertAuthChallengeHandler` still handles on device CBA the same, and now, as an example, it allows the device to detect whether or not a YubiKey is plugged in while the authentication WebView is shown.
Below is a summary of other notable file changes:

- `WebViewAuthorizationFragment`: A class variable was created to store the instance of `AzureActiveDIrectoryWebViewClient` so that it could be referenced later when the YubiKitManager needs to end Usb discovery.
- `AzureActiveDirectoryWebViewClient`: A new ~~`IChallengeHandler`~~ `ClientCertAuthChallengeHandler` class variable is created ~~and is set to either `EoClientCertAuthChallengeHandler` or `ClientCertAuthChallengeHandler` depending on the device's api level.~~ The `IChallengeHandler` is created in the constructor in order to ensure that the `YubiKitManager` (which is created in the `ClientCertAuthChallengeHandler`) has enough time to detect a plugged in YubiKey.
- When `WebViewAuthorizationFragment` is getting destroyed, it calls a wrapper method that stops the `YubiKitManager` from scanning for new Usb connections. Without this, the manager will continue to scan for Usb devices for the duration of the app session.

## Testing
Current tests in _Cert-Based Auth (TLS Challenge)_ in the manual test plans cover testing to make sure that on-device CBA still works with `ClientCertAuthChallengeHandler`. Detailed manual tests will be added to the same category once Smartcard CBA is fully complete.
When a YubiKey is plugged in, an alertDialog pops up that asks the user for permission to access the key. This is implemented by the YubiKit SDK. If the user accepts, the YubiKitManager now has access to the key. If the user denies, the YubiKitManager ignores the key.
To make sure that the YubiKey was being detected by the YubiKit SDK, I plugged in the YubiKey at various stages of the CBA flow while testing AzureSample, with and without BrokerHost:

- Plug in before clicking "Acquire Token"
    1. Plug in YubiKey.
    2. No logs that device was connected initially.
    3. Click "Acquire Token".
    4. Should be prompted by AlertDialog to authorize connection with YubiKey. Click "Accept".
    5. Log should show that device was connected.
    6. Finish on-device CBA.
    7. Upon successful login and return to AzureSample, log should show that device was disconnected.
- Plug in YubiKey after clicking "AcquireToken" (while WebView is showing)
    1. Click "Acquire Token".
    2. Once WebView is visible, plug in YubiKey.
    3. Should be prompted by AlertDialog to authorize connection with YubiKey. Click "Accept".
    4. Log should show that device was connected.
    5. Finish on-device CBA.
    6. Upon successful login and return to AzureSample, log should show that device was disconnected.
 - Plug in YubiKey after clicking "AcquireToken", and unplug while WebView is showing.
    1. Click "Acquire Token".
    2. Once WebView is visible, plug in YubiKey.
    3. Should be prompted by AlertDialog to authorize connection with YubiKey. Click "Accept".
    4. Log should show that device was connected.
    5. Unplug YubiKey.
    6. Log should show that device was disconnected.
    7. Finish on-device CBA.
    8. CBA should finish successfully with no crashes.

Another alternative flow that could be included in each of the scenarios above is denying connection with YubiKey in AlertDialog (No logs that show device is connected or disconnected). Due to the addition of the `keyboard` and `keyboardHidden` to Activity `configChanges` in the Manifests, none of the scenarios above should result in crashing. 

(Thank you Fadi and Dome for the testing advice :))
